File: 876-middle-of-the-linked-list/876-middle-of-the-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* middleNode(ListNode* head) {
        
        ListNode* curr = head;
        int length = 0;
        
        while (curr) {
            curr = curr->next;
            length++;
        }
        
        length = (length / 2) + 1;
        curr = head;
        
        while (--length > 0) {
            curr = curr->next;
        }
        
        return curr;
        
    }
};

File: 876-middle-of-the-linked-list/876-middle-of-the-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        middle = 0
        
        curr = head
        
        while (curr):
            curr = curr.next
            middle += 1

        curr = head
        middle = middle // 2
            
        for i in range(middle):
            curr = curr.next
        
        return curr
        